Compile error of high gui with QT

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

## System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => master
- Operating System / Platform => macOS
- Compiler => clang

## Error log
```
[ 76%] Built target opencv_videoio
[ 76%] Building CXX object modules/highgui/CMakeFiles/opencv_highgui.dir/src/window_QT.cpp.o
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:970:29: error: 
      no member named 'mapToGlobal' in 'ViewPort'
    QPoint org = w->myView->mapToGlobal(new QPoint(0, 0));
                 ~~~~~~~~~  ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:970:21: error: 
      'myView' is a private member of 'CvWindow'
    QPoint org = w->myView->mapToGlobal(new QPoint(0, 0));
                    ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.h:346:15: note: 
      declared private here
    ViewPort* myView;
              ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:22: error: 
      'myView' is a private member of 'CvWindow'
    return cvRect(w->myView->viewport()->pos().x() + org.x, w->myView->v...
                     ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.h:346:15: note: 
      declared private here
    ViewPort* myView;
              ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:58: error: 
      reference to non-static member function must be called; did you mean to
      call it with no arguments?
    return cvRect(w->myView->viewport()->pos().x() + org.x, w->myView->v...
                                                     ~~~~^
                                                          ()
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:64: error: 
      'myView' is a private member of 'CvWindow'
    return cvRect(w->myView->viewport()->pos().x() + org.x, w->myView->v...
                                                               ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.h:346:15: note: 
      declared private here
    ViewPort* myView;
              ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:100: error: 
      reference to non-static member function must be called; did you mean to
      call it with no arguments?
  ...+ org.x, w->myView->viewport()->pos().y() + org.y, w->myView->viewport()...
                                                 ~~~~^
                                                      ()
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:106: error: 
      'myView' is a private member of 'CvWindow'
  ...+ org.x, w->myView->viewport()->pos().y() + org.y, w->myView->viewport()...
                                                           ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.h:346:15: note: 
      declared private here
    ViewPort* myView;
              ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:138: error: 
      'myView' is a private member of 'CvWindow'
  ...+ org.y, w->myView->viewport()->width(), w->myView->viewport()->height());
                                                 ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.h:346:15: note: 
      declared private here
    ViewPort* myView;
              ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:30: error: 
      no member named 'viewport' in 'ViewPort'
    return cvRect(w->myView->viewport()->pos().x() + org.x, w->myView->v...
                  ~~~~~~~~~  ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:72: error: 
      no member named 'viewport' in 'ViewPort'
    return cvRect(w->myView->viewport()->pos().x() + org.x, w->myView->viewp...
                                                            ~~~~~~~~~  ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:114: error: 
      no member named 'viewport' in 'ViewPort'
  ...+ org.x, w->myView->viewport()->pos().y() + org.y, w->myView->viewport()...
                                                        ~~~~~~~~~  ^
/Users/xxx/software/opencv/modules/highgui/src/window_QT.cpp:976:146: error: 
      no member named 'viewport' in 'ViewPort'
  ...+ org.y, w->myView->viewport()->width(), w->myView->viewport()->height());
                                              ~~~~~~~~~  ^
12 errors generated.
make[2]: *** [modules/highgui/CMakeFiles/opencv_highgui.dir/src/window_QT.cpp.o] Error 1
make[1]: *** [modules/highgui/CMakeFiles/opencv_highgui.dir/all] Error 2
make: *** [all] Error 2
```

## Configuration
```.sh
#!/bin/bash

cmake \
 -D CMAKE_BUILD_TYPE=RELEASE \
 -D CMAKE_INSTALL_PREFIX=$HOME/software/opencv-install \
 \
 -D WITH_TBB=ON \
 -D BUILD_NEW_PYTHON_SUPPORT=OFF \
 -D WITH_V4L=OFF \
 -D INSTALL_C_EXAMPLES=OFF \
 -D INSTALL_PYTHON_EXAMPLES=OFF \
 -D BUILD_EXAMPLES=OFF \
 -D BUILD_DOCS=OFF \
 -D WITH_QT=ON \
 -D WITH_OPENGL=OFF \
 -D WITH_CUDA=OFF \
 -D WITH_OPENCL=OFF \
 -D WITH_MATLAB=OFF \
 -D WITH_VTK=ON \
 -DBUILD_TESTS=OFF \
 -DBUILD_PERF_TESTS=OFF \
 \
 -DBUILD_opencv_java=OFF \
 -DBUILD_python_bindings_generator=OFF \
 -DBUILD_opencv_python2=OFF \
 -DBUILD_opencv_python3=OFF \
 -DBUILD_opencv_ml=OFF \
 -DBUILD_opencv_objdetect=OFF \
 -DBUILD_opencv_superres=OFF \
 -DBUILD_opencv_videostab=OFF \
 \
 -DBUILD_opencv_phase_unwrapping=OFF \
 -DBUILD_opencv_plot=OFF \
 -DBUILD_opencv_reg=OFF \
 -DBUILD_opencv_surface_matching=OFF \
 -DBUILD_opencv_xphoto=OFF \
 -DBUILD_opencv_bgsegm=OFF \
 -DBUILD_opencv_face=OFF \
 -DBUILD_opencv_freetype=OFF \
 -DBUILD_opencv_fuzzy=OFF \
 -DBUILD_opencv_img_hash=OFF \
 -DBUILD_opencv_xobjdetect=OFF \
 -DBUILD_opencv_bioinspired=OFF \
 -DBUILD_opencv_dpm=OFF \
 -DBUILD_opencv_line_descriptor=OFF \
 -DBUILD_opencv_saliency=OFF \
 -DBUILD_opencv_text=OFF \
 -DBUILD_opencv_ccalib=OFF \
 -DBUILD_opencv_cvv=OFF \
 -DBUILD_opencv_datasets=OFF \
 -DBUILD_opencv_rgbd=OFF \
 -DBUILD_opencv_structured_light=OFF \
 -DBUILD_opencv_tracking=OFF \
 -DBUILD_opencv_videostab=OFF \
 -DBUILD_opencv_aruco=OFF \
 -DOPENCV_EXTRA_MODULES_PATH=../opencv_contrib/modules \
 ../opencv
```




<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->

--- a/modules/highgui/src/window_QT.cpp
+++ b/modules/highgui/src/window_QT.cpp
@@ -967,13 +967,8 @@ CvRect GuiReceiver::getWindowRect(QString name)
     if (!w)
         return cvRect(-1, -1, -1, -1);
 
-    QPoint org = w->myView->mapToGlobal(new QPoint(0, 0));
-#ifdef HAVE_QT_OPENGL
-    if (isOpenGl()) {
-        return cvRect(w->myView->pos().x() + org.x, w->myView->pos().y() + org.y, w->myView->width(), w->myView->height());
-    } else
-#endif
-    return cvRect(w->myView->viewport()->pos().x() + org.x, w->myView->viewport()->pos().y() + org.y, w->myView->viewport()->width(), w->myView->viewport()->height());
+    QPoint org = w->viewport()->mapToGlobal(new QPoint(0, 0));
+    return cvRect(w->viewport()->pos().x() + org.x(), w->viewport()->pos().y() + org.y(), w->viewport()->width(), w->viewport()->height());
 }
 
 double GuiReceiver::isFullScreen(QString name)

--- a/modules/highgui/src/window_QT.cpp
+++ b/modules/highgui/src/window_QT.cpp
@@ -1777,14 +1777,10 @@ void CvWindow::setRatio(int flags)
 
 CvRect CvWindow::getWindowRect()
 {
-    QPoint org = myView->viewport()->mapToGlobal(new QPoint(0, 0));
-
-#ifdef HAVE_QT_OPENGL
-    if (isOpenGl()) {
-        return cvRect(myView->pos().x() + org.x(), myView->pos().y() + org.y(), myView->width(), myView->height());
-    } else
-#endif
-    return cvRect(myView->viewport()->pos().x() + org.x(), myView->viewport()->pos().y() + org.y(), myView->viewport()->width(), myView->viewport()->height());
+    QWidget* view = myView->getWidget();
+    QRect local_rc = view->geometry(); // http://doc.qt.io/qt-5/application-windows.html#window-geometry
+    QPoint global_pos = /*view->*/mapToGlobal(QPoint(local_rc.x(), local_rc.y()));
+    return cvRect(global_pos.x(), global_pos.y(), local_rc.width(), local_rc.height());
 }
 
 int CvWindow::getPropWindow()

--- a/modules/highgui/src/window_QT.cpp
+++ b/modules/highgui/src/window_QT.cpp
@@ -966,9 +966,8 @@ CvRect GuiReceiver::getWindowRect(QString name)
 
     if (!w)
         return cvRect(-1, -1, -1, -1);
-
-    QPoint org = w->viewport()->mapToGlobal(new QPoint(0, 0));
-    return cvRect(w->viewport()->pos().x() + org.x(), w->viewport()->pos().y() + org.y(), w->viewport()->width(), w->viewport()->height());
+    
+    return w->getWindowRect();
 }
 
 double GuiReceiver::isFullScreen(QString name)
@@ -1776,6 +1775,17 @@ void CvWindow::setRatio(int flags)
     myView->setRatio(flags);
 }
 
+CvRect CvWindow::getWindowRect()
+{
+    QPoint org = myView->viewport()->mapToGlobal(new QPoint(0, 0));
+
+#ifdef HAVE_QT_OPENGL
+    if (isOpenGl()) {
+        return cvRect(myView->pos().x() + org.x(), myView->pos().y() + org.y(), myView->width(), myView->height());
+    } else 
+#endif
+    return cvRect(myView->viewport()->pos().x() + org.x(), myView->viewport()->pos().y() + org.y(), myView->viewport()->width(), myView->viewport()->height()); 
+}
 
 int CvWindow::getPropWindow()
 {

--- a/modules/highgui/src/window_QT.cpp
+++ b/modules/highgui/src/window_QT.cpp
@@ -966,7 +966,7 @@ CvRect GuiReceiver::getWindowRect(QString name)
 
     if (!w)
         return cvRect(-1, -1, -1, -1);
-    
+
     return w->getWindowRect();
 }
 
@@ -1782,9 +1782,9 @@ CvRect CvWindow::getWindowRect()
 #ifdef HAVE_QT_OPENGL
     if (isOpenGl()) {
         return cvRect(myView->pos().x() + org.x(), myView->pos().y() + org.y(), myView->width(), myView->height());
-    } else 
+    } else
 #endif
-    return cvRect(myView->viewport()->pos().x() + org.x(), myView->viewport()->pos().y() + org.y(), myView->viewport()->width(), myView->viewport()->height()); 
+    return cvRect(myView->viewport()->pos().x() + org.x(), myView->viewport()->pos().y() + org.y(), myView->viewport()->width(), myView->viewport()->height());
 }
 
 int CvWindow::getPropWindow()

--- a/modules/highgui/src/window_QT.h
+++ b/modules/highgui/src/window_QT.h
@@ -299,6 +299,7 @@ public:
     double getRatio();
     void setRatio(int flags);
 
+    CvRect getWindowRect();
     int getPropWindow();
     void setPropWindow(int flags);
 

--- a/modules/highgui/test/test_gui.cpp
+++ b/modules/highgui/test/test_gui.cpp
@@ -84,7 +84,8 @@ void CV_HighGuiOnlyGuiTest::run( int /*start_from */)
     waitKey(500);
 
     ts->printf(ts->LOG, "GUI 8\n");
-    getWindowImageRect("Win");
+    Rect rc = getWindowImageRect("Win");
+    std::cout << "window image rect: " << rc << std::endl;
 
     ts->printf(ts->LOG, "GUI 9\n");
     destroyAllWindows();

